[lldb][breakpoint] Grey out disabled breakpoints

chelcassanova opened the pull request:
This commit adds colour settings to the list of breakpoints in order to
grey out breakpoints that have been disabled.

bulbazord commented on lldb/include/lldb/Breakpoint/BreakpointOptions.h:
```diff
@@ -400,6 +400,12 @@ friend class Breakpoint;
+  Stream::HighlightSettings m_disbaled_breakpoint_highlight_settings{
+      "disabled",
+      "\x1b[31m",
+      "\x1b[37m",
```
You don't need to hardcode these values yourself. Take a look at `AnsiTerminal.h`, there are some constants defined for this purpose.

bulbazord commented on lldb/include/lldb/Breakpoint/BreakpointOptions.h:
```diff
@@ -400,6 +400,12 @@ friend class Breakpoint;
   /// Drawn from BreakpointOptions::SetOptionsFlags.
   Flags m_set_flags;
+  /// Settings that allow the 'disabled' keyword to be displayed in red.
+  Stream::HighlightSettings m_disbaled_breakpoint_highlight_settings{
```
typo: `disbaled` -> `disabled`

JDevlieghere commented on lldb/source/Breakpoint/BreakpointOptions.cpp:
```diff
@@ -534,7 +534,8 @@ void BreakpointOptions::GetDescription(Stream *s,
     if (m_ignore_count > 0)
       s->Printf("ignore: %d ", m_ignore_count);
-    s->Printf("%sabled ", m_enabled ? "en" : "dis");
+    s->PutCStringColorHighlighted(m_enabled ? "enabled " : "disabled ",
+                                  m_disbaled_breakpoint_highlight_settings);
```
This is going to regex search for disabled and then replace it with its colored variant. That seems unnecessarily inefficient. You know whether you're going to put enabled or disabled, so you could print it colored right away. 

You might also have to check the debugger setting for whether colors are enabled (`Debugger::GetUseColor()`). Sometimes it's set at the stream level: the stream might have colors disabled if it was created somewhere were the setting was checked). I don't know if that's always going to be the case here. If it's not, we should check the setting here. To find out I would run lldb with `--no-use-colors` and see if the colors still get printed.

chelcassanova replied:
Ah, so using `Printf` with `ansi::FormatAnsiTerminalCodes` instead? And yes using this with `--no-use-colors` does show the colour here so it looks like this isn't being checked at the stream level.

jimingham replied:
Note, you can get this output both when lldb decides to print in the terminal, and when someone does:

```
stream = lldb.SBStream()
bkpt.GetDescription(stream, eDescriptionLevelWhatever)

```

At present SBStream's are backed by StreamString objects which get constructed with use_color -> false.  So you should be okay there.

jimingham replied:
It seems odd you can't ask the stream whether it is using colors or not.  Because if I'm running with use-colors on, but I want to use the description of a breakpoint w/o colors - say to put it in a log - then I would correctly make an SBStream which is supposed to not use colors, and pass that to the description.  If you ONLY check whether the debugger is using colors, and override the stream's setting, then you will do the wrong thing.

chelcassanova replied:
This means that only the debugger tells you whether or not you're using colours then? `BreakpointOptions::GetDescription()` doesn't seem to have a reference to the debugger, target or interpreter so I'm wondering what the next best options to determine if we're using colours is.

jimingham replied:
This wouldn't be a problem if we make sure that whoever makes the Streams for use by the command interpreter consults GetUseColor when it creates the stream, and then we can rely on the stream to tell us what to do.  If you are calling stream API's that add the color for you, that should just work.  If you are doing a Print by hand with the color codes or not, you should consult the stream.  For that we'll have to add Stream::GetUseColor.  If we always get that right, the only problem will be if there are persistent streams, we may need to change their state or replace them when GetUseColor changes.

That seems the most principled way to solve this.

chelcassanova replied:
I was speaking to Ismail about this and seems like it would be nice if we had a `Stream::GetUseColor` for this given that you can set whether or not the stream has colours at construction but you can't query the state of it after that.

medismailben commented:
More of a styling preference, but I wouldn't use red as the color for disabled breakpoint, I think it would be more suited for unresolved ones (that could be done in a follow-up). Instead, I'd use the `ANSI_CTRL_FAINT` escape code to match what you can find in many IDEs.

chelcassanova committed:
[lldb][breakpoint] Grey out disabled breakpoints

This commit adds colour settings to the list of breakpoints in order to
grey out breakpoints that have been disabled.

chelcassanova commented:
I changed this PR to make disabled breakpoints greyed out (per Ismail's suggestion). Since this is being done at the `Breakpoint` level we have access to the target and its debugger to access `GetUseColor()`.

JDevlieghere commented on lldb/source/Breakpoint/Breakpoint.cpp:
```diff
@@ -848,6 +850,13 @@ void Breakpoint::GetDescription(Stream *s, lldb::DescriptionLevel level,
   const size_t num_resolved_locations = GetNumResolvedLocations();
 
+  // Grey out any disabled breakpoints in the list of breakpoints.
+  if (GetTarget().GetDebugger().GetUseColor())
+    s->Printf("%s",
+              IsEnabled()
+                  ? ansi::FormatAnsiTerminalCodes("${ansi.normal}").c_str()
+                  : ansi::FormatAnsiTerminalCodes("${ansi.faint}").c_str());
```
You need to reset the color after printing, otherwise everything in the terminal printed after will be faint.  Also seems like you can simplify this by moving the check for `IsEnabled()` into the first `if`:

```
const bool print_faint = !IsEnabled() && GetTarget().GetDebugger().GetUseColor();
if (print_faint)
     s->Print(ansi::FormatAnsiTerminalCodes("${ansi.faint}"));

[...]

if (print_faint)
     s->Print(ansi::FormatAnsiTerminalCodes("${ansi.reset}"));
```

chelcassanova replied:
![image](https://github.com/llvm/llvm-project/assets/21184907/0a95ccc6-bf9a-4acb-b8e0-2746d5911f15)
I also thought that I'd have to reset the colours back to normal somewhere but when I tried this for a little the colours seemed fine without explicitly placing a reset somewhere.

I can still add a reset alongside this simplified logic though.

jimingham commented:
I'm not sure checking the debugger here resolves the difficulty I pointed out earlier.  If I make an SBStream and call SBBreakpoint::GetDescription, I will have passed in an SBStream with `use_color` explicitly off.  It still seems to me that should take precedence over the debugger's global setting.  I wanted this for programmatic uses presumably, and shouldn't have to deal with color codes...  Having to temporarily unset the global use color in the debugger to get this effect is not a good design.

chelcassanova commented:
Hm, so in that case should we focus on adding an `SBStream::GetUseColor` so that the stream's colour settings can take precedence here?

jimingham commented:
> Hm, so in that case should we focus on adding an `SBStream::GetUseColor` so that the stream's colour settings can take precedence here?

That's the only way it makes sense to me.  But if we are going to rely on the stream, we also have to be sure that we're setting the Streams "use color" correctly when we make them.

That might be an annoying accounting exercise.

I think an easier way allow streams to both get their "use color" from context and also override it for special purposes is to replace the use color bool with a "use color, don't use color, no opinion" enum.  We'd make streams creation default to "no opinion".  The SB API's would instead create them with "no use color" - and we should add an accessor to the SB API's.  And then internally, "no opinion" means "consult the debugger" - but more generally it means "determine my use color value from context".

That saves you from having to track down all the cases where streams were made and get their "use color" right.

chelcassanova commented:
Cool, sorry for the delay but I made changes to `Stream` and `SBStream` to add a `GetUseColor()` accessor and to change the bool it uses in initialization to a `std::optional` and I'm gonna put them up in a separate PR.

JDevlieghere commented:
The stream already knows whether color support is enabled, but it looks like we're only storing that in the underlying llvm stream, which is commonly used in combination with `WithColor`. I think we could add a member to check if the stream has colors enabled. It's already a property of `m_forwarder` so we don't need to create a new member. 

```
SBStream::HasColor() {
  return m_forwarded.has_color();
}
```

Alternatively, we could add a helper that does that under the hood. 

```
SBStream::FormatAnsiTerminalCodes(llvm::StringRef format) {
  if (m_forwarded.has_color()) 
    print(FormatAnsiTerminalCodes(format));
}
```

So instead of having to write:

```
if(s->HasColor())
 s->Printf("%s", ansi::FormatAnsiTerminalCodes("${ansi.faint}").c_str());
```

You can just write:

```
s-> FormatAnsiTerminalCodes("${ansi.faint}") 
```

and have it do the right thing.
```

I think I understand Jim's concern. Most places that we use colors, we do that through `WithColor` which checks the color property of the underlying LLVM stream. Chelsea is not going through WithColor, so in her case she would indeed need to check

JDevlieghere reviewed (commented):
This is looking good, but needs a (pexpect) test. I would also like to add a setting to make the prefix and suffix configurable, like we do for things like `show-progress-ansi-prefix` and `show-autosuggestion-ansi-prefix`. That way user can pick their own preferred color, or even just turn this one feature off without having to disable color if they want to.

chelcassanova commented:
Just making sure, a setting for changing the prefix/suffix would be in `CoreProperties.td` right?

labath commented on lldb/source/API/SBStream.cpp:
```diff
@@ -77,6 +78,15 @@ void SBStream::Printf(const char *format, ...) {
+
+void SBStream::FormatAnsiTerminalCodes(llvm::StringRef format) {
+  if (HasColor())
+    Printf("%s", ansi::FormatAnsiTerminalCodes(format).c_str());
```
```suggestion
    m_opaque_up->PutCString(ansi::FormatAnsiTerminalCodes(format));
```

labath commented on lldb/test/API/terminal/TestDisabledBreakpoints.py:
```diff
@@ -0,0 +1,32 @@
+        """Test that disabling a breakpoint and viewing the breakpoints list uses the specified ANSI color prefix."""
+        import pexpect
+
+        self.child = pexpect.spawn("expect", encoding="utf-8")
+
```
???

chelcassanova replied:
This line is probably unnecessary, I was modeling this test after another `pexpect` test that spawns pexpect with the `expect` program but I don't think this tests needs that now that I'm looking at this again, I can remove this.

labath replied:
I'm pretty sure this was doing nothing as `self.launch` overwrites the `child` member. I recall we have one test which uses `pexpect` (the python module) to drive `expect` (the unix utility which lent its name to the former) to drive `lldb`. We definitely don't need that here (and tbh, I'm not sure the original test does either).

labath commented on lldb/test/API/terminal/TestDisabledBreakpoints.py:
```diff
@@ -0,0 +1,32 @@
+        ansi_red_color_code = "\x1b[31m"
+
+        self.launch(use_colors=True, dimensions=(100, 100))
+        self.child.sendline(
```
Any reason you can't use self.expect instead?

Even if you don't pass it any expectations to it, that function will at least check for the lldb prompt, which means it will wait for lldb to process the previous command before you feed it the next one. That makes the test much more deterministic and reliable.

labath replied:
Yes, that's what I meant. The `expect` function was meant to be used to inject lldb commands, some tests use the pexpect API directly because they're doing something more complicated (or because the thing they're driving is not lldb), but that's not the case here.

labath commented on lldb/test/API/terminal/TestDisabledBreakpoints.py:
```diff
@@ -0,0 +1,26 @@
+        self.expect('settings set disable-breakpoint-ansi-prefix "${ansi.fg.red}"')
+        self.expect("b main")
+        self.expect("br dis")
+        self.expect("br l")
+        self.child.expect_exact(ansi_red_color_code + "1:")
```
```suggestion
        self.expect("br l", substrs=[ansi_red_color_code + "1:"])
```

The idea is to make this mostly look like the existing SB API tests.

Kinda surprised it works without it. If I had to guess, I'd say its because of the prompt coloring code causes the prompt substring to be present more than once (which then means the substring search loses synch).

(That's one more reason to add some sort of expectations to the previous commands. The other reason is that they give you an earlier indication if something goes wrong during the test -- the pexpect error messages aren't of the most readable type)

chelcassanova replied:
Thanks for the suggestion to use `substrs` here! By "expectations" you just mean adding a `self.expect` after all the previous commands to check that their output is fine right? If so then yeah I can add that 👍🏾

labath replied:
What I meant was adding a `substr` argument to the existing `expect` calls (e.g. `self.expect("br dis", substrs=["All breakpoints disabled."])`), not adding new expect calls.

labath reviewed (commented):
I'm kinda late to the party (particularly as there's an analogous "progress" setting), but I have to say I was surprised to see a `disable-breakpoint-ansi-suffix` setting. That sounds like too low level of a model. The model I would expect/find more natural is if there was a single setting describing the "normal" text colour, and then several (not necessarily one each item that needs colouring) settings defining the colour of items that need colouring.

I find it hard to imagine why one would ever want to set different values for the "suffix" colours, as that means the colour of random text would depend on what was printed before it.

I don't want to make this a blocking objection, but I would urge you to consider if this is the behavior you want.

chelcassanova commented:
I also agree with how I was misusing the `pexpect` module here, I was way overcomplicating things there by trying to model my test after the one that drives the `expect` program instead of lldb itself 😅

JDevlieghere reviewed (changes requested):
I like the feautre but I think this PR can be a lot simpler:

1. I don't think you need any of the changes to `Stream` or `SBStream`. You already have to get the setting from the debugger, so just check if it has colors enabled and write the prefix (and fixed suffix) if colors are enabled. 
2. I think we had settled on adding a new setting for highlighting "disabled things", with breakpoints being just one example of that. The setting should be generic, something like `disable-ansi-prefix` (*). I'm find with omitting the suffix and always issuing a clear. I can't think of a reason anyone would ever want to change that.

(*) I would prefer to call the setting`disable-format`, but then it should go through FormatEntity which is definitely overkill. I think given the existing settings, `disable-ansi-prefix` is the most consistent.

chelcassanova commented:
Just updated this patch to:
- Remove changes to `SBStream` and `Stream`
- Change the setting name to `disable-ansi-prefix`

JDevlieghere commented on lldb/source/Breakpoint/Breakpoint.cpp:
```diff
@@ -934,6 +943,12 @@ void Breakpoint::GetDescription(Stream *s, lldb::DescriptionLevel level,
     s->IndentLess();
   }
+
+  // Reset the colors back to normal if they were previously greyed out.
+  if (s->AsRawOstream().colors_enabled())
```
The comment says "if they were previously greyed out" but you're not actually checking that. I think with the helper variable the code is sufficiently self-explanatory that we can drop the comment. 
```suggestion
  // Reset the colors back to normal if they were previously greyed out.
  if (highlight_disabled)
```

bulbazord commented on lldb/source/Breakpoint/Breakpoint.cpp:
```diff
@@ -838,6 +840,14 @@ void Breakpoint::GetDescription(Stream *s, lldb::DescriptionLevel level,
   assert(s != nullptr);
 
+  // Grey out any disabled breakpoints in the list of breakpoints.
+  const bool highlight_disabled =
```
Suggestion: rename this variable to "dim_breakpoint_description" or something like that.
The word "highlight_disabled" could also be applied to situations where colors are not enabled.

bulbazord commented on lldb/source/Breakpoint/Breakpoint.cpp:
```diff
@@ -838,6 +840,14 @@ void Breakpoint::GetDescription(Stream *s, lldb::DescriptionLevel level,
                                 bool show_locations) {
   assert(s != nullptr);
 
+  // Grey out any disabled breakpoints in the list of breakpoints.
```
nit: I suggest removing the comment. I think naming the conditions variable better should make this clearer.

chelcassanova replied:
Just applied both of these suggestions to the patch.
